api/turing/batch/controller/ensembling.go

```diff
@@ -0,0 +1,268 @@
+	if err != nil {
+		return SparkApplicationStateUnknown, fmt.Errorf("failed to retrieve spark application %v", err)
+	}
+
+	state := sa.Status.AppState.State
+	if state == apisparkv1beta2.CompletedState {
+		return SparkApplicationStateCompleted, nil
+	}
+
+	if state == apisparkv1beta2.FailedState {
+		return SparkApplicationStateFailed, nil
+	}
+
+	if state == apisparkv1beta2.UnknownState {
+		return SparkApplicationStateUnknown, nil
+	}
+
+	return SparkApplicationStateRunning, nil
+}
+
+func (c *ensemblingController) Create(request *CreateEnsemblingJobRequest) error {
+	var err error
+	defer func() {
+		if err != nil {
+			c.cleanup(request.EnsemblingJob.Name, request.Namespace)
+		}
+	}()
+
+	err = c.clusterController.CreateNamespace(request.Namespace)
+	if errors.Is(err, cluster.ErrNamespaceAlreadyExists) {
+		// This error is ok to ignore because we just need the namespace.
+		err = nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("failed creating namespace %s: %v", request.Namespace, err)
+	}
+
+	serviceAccount, err := c.createSparkDriverAuthorization(request.Namespace)
+	if err != nil {
+		return fmt.Errorf(
+			"failed creating spark driver authorization in namespace %s: %v",
+			request.Namespace,
+			err,
+		)
+	}
+
+	secretString, err := c.mlpService.GetSecret(
+		request.EnsemblingJob.ProjectID,
+		request.EnsemblingJob.InfraConfig.ServiceAccountName,
+	)
+	if err != nil {
+		return fmt.Errorf(
+			"service account %s is not found within %s project: %s",
+			request.EnsemblingJob.InfraConfig.ServiceAccountName,
+			request.Namespace,
+			err,
+		)
+	}
+
+	err = c.createSecret(request, secretString)
+	if err != nil {
+		return fmt.Errorf(
+			"failed creating secret for job %s in namespace %s: %v",
+			request.EnsemblingJob.Name,
+			request.Namespace,
+			err,
+		)
+	}
+
+	err = c.createJobConfigMap(request.EnsemblingJob, request.Namespace)
+	if err != nil {
+		return fmt.Errorf(
+			"failed creating job specification configmap for job %s in namespace %s: %v",
+			request.EnsemblingJob.Name,
+			request.Namespace,
+			err,
+		)
+	}
+
+	_, err = c.createSparkApplication(request, serviceAccount)
+	if err != nil {
+		return fmt.Errorf(
+			"failed submitting spark application to spark controller for job %s in namespace %s: %v",
+			request.EnsemblingJob.Name,
+			request.Namespace,
+			err,
+		)
+	}
+
+	return nil
+}
+
+func (c *ensemblingController) createSparkApplication(
+	jobRequest *CreateEnsemblingJobRequest,
+	serviceAccount *apicorev1.ServiceAccount,
+) (*apisparkv1beta2.SparkApplication, error) {
+	infraConfig := jobRequest.EnsemblingJob.InfraConfig
+	request := &cluster.CreateSparkRequest{
+		JobName:            jobRequest.EnsemblingJob.Name,
+		JobLabels:          jobRequest.Labels,
+		JobImageRef:        jobRequest.ImageRef,
+		JobApplicationPath: ensemblingJobApplicationPath,
+		JobArguments: []string{
+			"--job-spec",
+			fmt.Sprintf("%s%s", cluster.JobConfigMount, cluster.JobConfigFileName),
```

Comment:
Sorry I'm not sure if I fully get what you mean. Are you saying that it should be in the batch package rather than the batchcontroller package? Somewhat connected to the repackaing of packages? https://github.com/gojek/turing/pull/61#discussion_r659481407

Reply:
I'm saying, that your variables `JobConfigMount` and `JobConfigFileName` are defined under `cluster` package (https://github.com/ashwinath/turing/blob/ae4317409c67f41eca04c402831368dfdece8013/api/turing/cluster/spark.go#L19), but I'm suggesting to move them under `api/turing/batch` package to incapsulate all batch related stuff in a single package. I also added a link to Merlin, to give you an idea, how Merlin folks structured their variables related to batch processing.

Reply:
Ah! Ok must be having some brain fart when I wrote this. Yah, I get what you mean now. Will change this.